src/binary_tree.rs

```diff
@@ -0,0 +1,380 @@
+//! A safe binary tree.
+//!
+//! The `binary tree` allows inserting, removing and is naturally sorted.
+//!
+//! NOTE: This was written for a learning purpose.
+use super::stack::Stack;
+use std::cmp::Ordering;
+
+/// A binary tree build from Nodes. This struct represents a binary tree
+/// holding a root node.
+pub struct BinaryTree<T> {
+    head: Link<T>,
+}
+
+/// A Link between Nodes.
+type Link<T> = Option<Box<Node<T>>>;
+
+/// A Node in a binary tree which holds a reference to the left and right Nodes as well as a value.
+#[derive(Debug, Eq, PartialEq)]
+struct Node<T> {
+    left: Link<T>,
+    right: Link<T>,
+    value: T,
+}
+
+/// The Iterator for a binary tree, containing a stack with all nodes that should be visited.
+/// Instances are created by [`BinaryTree::iter()`]. See its
+/// documentation for more.
+pub struct Iter<'a, T: 'a> {
+    visited: Box<Stack<&'a Box<Node<T>>>>,
+}
+
+impl<T: Eq + std::cmp::Ord> BinaryTree<T> {
+    /// Creates a new and empty `BinaryTree`.
+    /// # Example
+    /// ```rust
+    /// use data_structure_with_colin::binary_tree::BinaryTree;
+    /// let mut binary_tree = BinaryTree::<()>::new();
+    /// assert!(binary_tree.is_empty());
+    ///```
+    pub fn new() -> Self {
+        BinaryTree { head: None }
+    }
+
+    /// Inserts a new element into the tree. The tree alway keeps an ordered structure by
+    /// making sure that the left child node is always "bigger" than the right child node.
+    /// # Example
+    /// ```rust
+    /// use data_structure_with_colin::binary_tree::BinaryTree;
+    /// let mut binary_tree = BinaryTree::new();
+    /// binary_tree.insert(1);
+    /// binary_tree.insert(2);
+    ///
+    /// assert!(binary_tree.find(1));
+    /// assert!(binary_tree.find(2));
+    /// ```
+    pub fn insert(&mut self, value: T) -> bool {
+        if self.is_empty() {
+            self.head = Some(Box::new(Node::new(value)));
+            return true;
+        }
+        self.head.as_mut().unwrap().insert(value)
+    }
+
+    /// Checks if the tree is empty.
+    pub fn is_empty(&self) -> bool {
+        self.head.is_none()
+    }
+
+    /// Searches for an element in the tree and returns whether the tree contains the element
+    /// or not. As the tree is always ordered, this happens relatively fast. (As long as it's not)
+    /// super unordered.
+    /// # Example
+    /// ```rust
+    /// use data_structure_with_colin::binary_tree::BinaryTree;
+    /// let mut binary_tree = BinaryTree::new();
+    /// binary_tree.insert(1);
+    ///
+    /// assert!(binary_tree.find(1));
+    /// ```
+    pub fn find(&self, value: T) -> bool {
```

Comment:
pls rename to contains, find implies you return the value (or do that and keep the name)